File: bench/benches/hash.lua
-- mutable hash-table throughput (see bench/benches/hash.l). checksum = N*N.
-- keys are sparse (stride 97) so they land in the table's hash part, not the
-- contiguous array part.
package.path = (arg[0]:match("(.*/)") or "./") .. "../lib/?.lua;" .. package.path
local bench = require("bench")
local N = 10000
bench("hash", function()
  local h = {}
  for i = 0, N - 1 do h[97 * i + 1] = i end
  local a = 0
  for i = 0, N - 1 do a = a + h[97 * i + 1] end
  for i = 0, N - 1 do local k = 97 * i + 1; h[k] = h[k] + 1 end
  for i = 0, N - 1 do a = a + h[97 * i + 1] end
  return a
end)

File: bench/benches/hash.java
import java.util.HashMap;

// mutable hash-table throughput. checksum = N*N.
class Main {
    static final int N = 10000;

    static long work() {
        HashMap<Integer, Integer> h = new HashMap<>();
        for (int i = 0; i < N; i++) {
            h.put(97 * i + 1, i);
        }
        long acc = 0;
        for (int i = 0; i < N; i++) {
            acc += h.get(97 * i + 1);
        }
        for (int i = 0; i < N; i++) {
            int k = 97 * i + 1;
            h.put(k, h.get(k) + 1);
        }
        for (int i = 0; i < N; i++) {
            acc += h.get(97 * i + 1);
        }
        return acc;
    }

    public static void main(String[] a) {
        Bench.bench("hash", Main::work);
    }
}
